modules/lightmapper_rd/lm_compute.glsl

```diff
@@ -1037,4 +1038,28 @@ void main() {
 		imageStore(dest_light, ivec3(atlas_pos, lightmap_slice), vec4(denoised_rgb, input_light.a));
 	}
 #endif
+
+#ifdef MODE_PACK_L1_COEFFS
+	vec4 base_coeff = texelFetch(sampler2DArray(source_light, linear_sampler), ivec3(atlas_pos, params.atlas_slice * 4), 0);
+
+	for (int i = 1; i < 4; i++) {
+		vec4 c = texelFetch(sampler2DArray(source_light, linear_sampler), ivec3(atlas_pos, params.atlas_slice * 4 + i), 0);
+
+		if (abs(base_coeff.r) > 0.0) {
+			c.r /= (base_coeff.r * 8);
```

Comment:
Where is the 8 coming from? From the presentation I remember that they factored out the 2 and then post-multiplied that in the shader. I understand we are combining the [-1,1] -> [0,1] into one step, but then it should be 4 here and 2 in the reading shader. 

We are dividing and multiplying by an extra factor of 2 here as far as I can tell

Reply:
In the paper the L1 data is also multiplied by 0.5 at bake time. Godot doesn't do that, so we have to include the extra '2' in the division.

Reply:
In the paper, this is mentioned under the diffuse evaluation:
![aa](https://github.com/user-attachments/assets/8cad4d44-ff57-4d03-930f-dd31e54453a9)

I might be misunderstanding this, but in the paper the L1 data is 2x as big as L0, but it needs to be smaller or equal, hence the first division. Then it's in the [-1 - 1] range, so it's divided again and 0.5 is added to place it in the [0 - 1] range. In summary, the total value they need to be divided by is 4.

 In Godot, the L1 harmonics are 4x bigger than L0, which needs to be reflected when packing. The first division by 4 brings the data into the [-1 - 1] range, and another by 2 (and +0.5) into [0 - 1], hence the 8

Reply:
We use the same coefficients as the paper.

In the paper they multiply by two outside of the "EvaluateDiffuseL1()" function. In Godot I factored out the multiplication by 2 and the multiplication by 0.5 which is why our L1 coefficients are seemingly multiplied by nothing.

Reply:
Whatever you end up doing please document this sort of stuff for future contributors so that people don't end up throwing things off when trying to optimize

Reply:
> Whatever you end up doing please document this sort of stuff for future contributors so that people don't end up throwing things off when trying to optimize

This is the second time you have suggested this despite the code already having extensive comments: https://github.com/godotengine/godot/blob/88197d4a513a7873cb7cf85138a42a04fb1c9011/modules/lightmapper_rd/lm_compute.glsl#L652-L655

The problem is this stuff is super confusing and math heavy. So even with a detailed comment it hard to get things right. As you can see above both BlueCube and I both misunderstood aspects of how the SH should be encoded despite having comments

Reply:
Github commit view can get weird in large PRs. Totally missed some of the comments that already exist :/

Reply:
> edit: as I go through the presentation again, I see that the coefficients we use are wrong. The 2 times multiplier is what is left after multiplying the 0.5 coefficient by the 4 * PI from the Monte Carlo estimation. Similarly L0 should be multiplied by 1 (the PI is factored out by the BRDF).

On second thought, the results look fine now. Also, for direct lighting we don't do a Monte Carlo estimation. It seems we also already bake in the `1/PI` from the BRDF. So my math is probably totally wrong but somehow arriving at something similar to the correct values. I have a feeling that we should be integrating over a hemisphere rather than a sphere. I'll check

Reply:
I'm making some progress on this! Indeed the 4 * PR factor should not be used for the direct pass. For the indirect pass, it should be 2 * PI, but we don't weight the samples, so no normalization is needed there. 

However, https://www.ppsloan.org/publications/StupidSH36.pdf (at page 12) indicates that the normalization factor for Directional Lights is `16 * PI / 17`. I'm about to try recalculating the coefficients but replacing `4 * PI` with `16 * PI / 17` and removing the BRDF's `1 / PR` completely

Reply:
> Also, as I research this, I am finding that the way we bake static lighting for omnilights and spotlights is just completely wrong. We are essentially approximating them as if they were a directionalight.

@clayjohn that will change the way they look?

Reply:
> > Also, as I research this, I am finding that the way we bake static lighting for omnilights and spotlights is just completely wrong. We are essentially approximating them as if they were a directionalight.
> 
> @clayjohn that will change the way they look?

Yes, but I think the difference will be quite subtle. Basically they will appear softer than they should which is probably fine in most cases. I'm not going to worry about it unless someone tells me their Omni/Spot lights look horrible

Reply:
@BlueCube3310 I've discovered enough that I realize its probably best to maintain the division by 8 here and just live with this small hack for now. 

I am now totally certain that we are baking the spherical harmonics for direct lighting wrong. However, things look fine in most cases, so I am happy to live with the problems until we fix things properly. it doesn't make sense for this PR to wait until Directional baking is fully fixed. 

In other words, I think this is the correct code given the problems that exist for the directional baking right now.